src/main/java/com/aws/greengrass/clientdevices/auth/certificate/infra/BackgroundCertificateRefresh.java

```diff
@@ -104,18 +129,144 @@ public void run() {
-        String[] certificateIds = registry.getAllCertificateIds();
+        ThingAssociations associations = getThingsAssociatedWithCoreDevice();
+        refresh(associations);
+    }
+
+    /**
+     * Returns ThingAssociations which has information about what Things are associated with the core device and
+     * which things are no longer associated with the core device.
+     */
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private ThingAssociations getThingsAssociatedWithCoreDevice() {
+        ThingAssociations associations = new ThingAssociations();
+        RetryUtils.RetryConfig retryConfig = RetryUtils.RetryConfig.builder()
+                .initialRetryInterval(Duration.ofSeconds(30)).maxRetryInterval(Duration.ofMinutes(3))
+                .maxAttempt(3)
+                .retryableExceptions(Arrays.asList(ThrottlingException.class, InternalServerException.class)).build();
+
+        try {
+            Set<String> cloudAttachedThingNames =  RetryUtils.runWithRetry(
+                    retryConfig, iotAuthClient::getThingsAssociatedWithCoreDevice,
+                            "get-things-associated-with-core-device", logger)
+                    .map(Thing::getThingName).collect(Collectors.toSet());
+
+            thingRegistry.getAllThings().forEach(thing -> {
+                if (cloudAttachedThingNames.contains(thing.getThingName()))  {
+                    associations.attached(thing);
+                } else {
+                    associations.detached(thing);
+                }
+            });
+        } catch (AccessDeniedException e) {
+            logger.atWarn().log(
+                    "Access denied to list things associated with core device. Please make sure you "
+                    + "have setup the correct permissions.");
+        } catch (Exception e) {
+            logger.warn("Failed to get things associated to the core device. Retry will be schedule later");
+        }
+
+        return associations;
+    }
 
-        for (String certificateId : certificateIds) {
-            if (pemStore.exists(certificateId)) {
-                String pem = pemStore.getPem(certificateId).get();
-                verifyIotCertificate.apply(pem);
+    private void refresh(ThingAssociations associations) {
+       associations.getThingAssociations().forEach(this::refreshCertificateThingAttachment);
+       associations.getCertificateAssociations().forEach(this::refreshCertificateValidity);
+       associations.getStaleCertificateAssociations().forEach(certificateRegistry::deleteCertificate);
+       associations.getStaleThingAssociations().forEach(thingRegistry::deleteThing);
+    }
+
+    private void refreshCertificateThingAttachment(Thing thing) {
+        VerifyThingAttachedToCertificate useCase = useCases.get(VerifyThingAttachedToCertificate.class);
+
+        thing.getAttachedCertificateIds().keySet().forEach(certificateId -> {
+            Optional<String> certPem = pemStore.getPem(certificateId);
```

Comment:
We don't need PEM for verifying Thing <-> Certificate attachment. Just CertId and thing name would do. How about updating `VerifyThingAttachedToCertificateDTO` to have only thingName and CertId instead of complete domain objects?

Reply:
Yeah it was just that we don't expose creating a certificate without the pem and if for some reason we don't have the pem stored and we can't update the cert validity would we still want to update its attachements?